### PR TITLE
Enhancement/issue 590 improve the fault tolerance of materialized view

### LINKED ISSUE #590: improve the fault tolerance of materialized view
**Describe what enhancement you'd like to have**
When send dirty data to proton (for example, an integer to a float32 column)
it will cause the mv failed with exception

proton will try to recover the fail mv forever. But due to the dirty data, it cannot recover successfully.

Expected Result:
proton can log the dirty data which cause the failure
when recover mv, proton can skip the dirty and recover the mv successfully.

### LINKED ISSUE #598: Poison event caused MV to be unable to progress and recover
**Describe what enhancement you'd like to have**

Typical streaming ETL like json extract or regex, if there is dirty data (poison event) in the stream, MV will end up infinite loop since for now it can't skip these dirty messages and the error message of the MV is hard to consume as well.

We will need find a way to do
1. Skip these dirty messages
2. Do appropriate logging
3. Show more readable error message : best with raw data as the context in the error message

### PR DESCRIPTION
**Pull request by yl-lisen**
resolve #590 and close #598

Please write user-readable short description of the changes:
- Add an abstract class `DB::Streaming::ISource`
- Support more recovery policy `RecoveryPolicy` via a setting `settings recovery_policy='strict' or 'best_effort'`:
  ```
  - Strict      /// Always recover from checkpointed
  - BestEffort  /// Attempt to recover from checkpointed and allow skipping of some data with permanent errors
  ```
- Add more information in error log (include the description for streaming source and sequence number (such as progressed sn, checkpointed sn, recovered sn)

If we need know what's the dirty data, we can use the following statement to query data within the specified sn range `(checkpointed_sn, processed_sn]`:
```sql
select * from source_stream where _tp_sn > checkpointed_sn and _tp_sn <= processed_sn
-- or
select * from source_mv where _tp_sn > checkpointed_sn and _tp_sn <= processed_sn
```
Usually, the data at `processed_sn` is the cause of the error

### CODE REVIEW

File: src/Core/RecoveryPolicy.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <Core/SettingsFields.h>
+
+namespace DB
+{
+enum class RecoveryPolicy : uint8_t
+{
+    Strict = 0,      /// Always recover from checkpointed
+    // (TODO) Cautious = 1,   /// Attempt to recover from checkpointed and allow skipping of specific data that would cause a permanent error
```
Comment (chenziliang) [inline src/Core/RecoveryPolicy.h]: Let's remove `=1`

File: src/Core/RecoveryPolicy.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    Loose = 2,     /// Attempt to recover from checkpointed and allow skipping of some data with permanent errors
```
Comment (chenziliang) [inline src/Core/RecoveryPolicy.h]: `BestEffort` ?

File: src/Processors/Streaming/ISource.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#include <Processors/Streaming/ISource.h>
+
+namespace DB::Streaming
+{
+/// It basically initiate a checkpoint
+/// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
+/// We nee make sure it is thread safe
```
Comment (chenziliang) [inline src/Processors/Streaming/ISource.cpp]: `need `

File: src/Processors/Streaming/ISource.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+public:
+    ISource(Block header, bool enable_auto_progress, ProcessorID pid_) : DB::ISource(std::move(header), enable_auto_progress, pid_)
+    {
+        is_streaming = true;
+    }
+
+    virtual String description() const { return ""; }
+
+    /// \brief Get the last progressed sequence number of the source, it shouldn't be called during pipeline execution (thread-unsafe)
+    virtual Int64 lastSN() const { throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented for lastSN() of {}", getName()); }
```
Comment (chenziliang) [inline src/Processors/Streaming/ISource.h]: `lastProcessedSN`, `lastCheckpointedSN`, `resetStartSN()`

File: src/Storages/ExternalStream/Kafka/KafkaSource.cpp [hunk 5]
```diff
@@ -335,16 +330,6 @@ void KafkaSource::calculateColumnPositions()
...
 }
 
-
-/// It basically initiate a checkpoint
-/// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
-/// We nee make sure it is thread safe
-void KafkaSource::checkpoint(CheckpointContextPtr ckpt_ctx_)
-{
-    /// We assume the previous ckpt is already done
-    ckpt_request.setCheckpointRequestCtx(ckpt_ctx_);
```
Comment (chenziliang) [inline src/Storages/ExternalStream/Kafka/KafkaSource.cpp:345]: do we need delete `ckpt_request` in this class at line 96 ?

File: src/Processors/Streaming/ISource.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    /// \brief Reset the start sequence number of the source, it must be called before the pipeline execution (thread-unsafe)
+    void resetStartSN(Int64 sn);
+
+    Int64 lastCheckpointedSN() const noexcept { return last_checkpointed_sn; }
+
+    void checkpoint(CheckpointContextPtr ckpt_ctx_) override final;
```
Comment (chenziliang) [inline src/Processors/Streaming/ISource.h:34]: can we make `checkpoint` non virtual and remove the `checkpoint, recover` interface from DB::ISource?

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 7]
```diff
@@ -343,15 +389,19 @@ void StorageMaterializedView::initBackgroundState()
 
             /// FIXME: limit retry times ?
             size_t retry_times = 0;
+            /// Keep the last SNs of the stream source during the last retries.
+            std::deque<std::vector<Int64>> failed_sns_queue;
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: `failed_sn_queue`

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 8]
```diff
@@ -73,6 +74,49 @@ String generateInnerQueryID(const StorageID & view_id)
         return ".inner.query-id.from-" + toString(view_id.uuid);
     return ".inner.query-id.from-" + view_id.getTableName();
 }
+
+void resetSNsOfStreamingSources(const auto & streaming_sources, const std::vector<Int64> & sns)
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: `resetSNForStreamingSources`

Comment (chenziliang) [reply src/Storages/Streaming/StorageMaterializedView.cpp]: for input param, let's avoid using `auto` for now

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 9]
```diff
@@ -73,6 +74,49 @@ String generateInnerQueryID(const StorageID & view_id)
...
+{
+    if (sns.size() != streaming_sources.size()) [[unlikely]]
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Number of sequence numbers doesn't match number of streaming sources");
+
+    for (auto [streaming_source, sn] : std::views::zip(streaming_sources, sns))
+        streaming_source->resetStartSN(sn);
+}
+
+std::vector<Int64> getSNsOfStreamingSources(const auto & streaming_sources)
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: `getSNOfStreamingSources`

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 10]
```diff
@@ -73,6 +74,49 @@ String generateInnerQueryID(const StorageID & view_id)
...
+
+    if (sns.empty())
+    {
+        for (const auto & streaming_source : streaming_sources)
+            wb << streaming_source->getName() << "(" << streaming_source->description() << ",sn=" << streaming_source->lastCheckpointedSN()
+               << ");";
+    }
+    else
+    {
+        if (sns.size() != streaming_sources.size()) [[unlikely]]
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: let's move unlikely, it is exception code path anyway

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 11]
```diff
@@ -73,6 +74,49 @@ String generateInnerQueryID(const StorageID & view_id)
         return ".inner.query-id.from-" + toString(view_id.uuid);
     return ".inner.query-id.from-" + view_id.getTableName();
 }
+
+void resetSNsOfStreamingSources(const auto & streaming_sources, const std::vector<Int64> & sns)
+{
+    if (sns.size() != streaming_sources.size()) [[unlikely]]
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Number of sequence numbers doesn't match number of streaming sources");
+
+    for (auto [streaming_source, sn] : std::views::zip(streaming_sources, sns))
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: `auto &` ?

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 12]
```diff
@@ -417,23 +470,64 @@ void StorageMaterializedView::initBackgroundState()
...
-
-                        /// For some queries that always go wrong, we don’t want to recover too frequently. Now wait 5s, 10s, 15s, ... 30s, 30s for each time
-                        std::this_thread::sleep_for(recover_interval * std::min<size_t>(retry_times, 6));
+                        auto streaming_sources = io.pipeline.getStreamingSources();
+                        switch (local_context->getSettingsRef().recovery_policy)
+                        {
+                            case RecoveryPolicy::Strict: {
+                                background_state.setException(
+                                    err_code,
+                                    fmt::format(
```
Comment (chenziliang) [inline src/Storages/Streaming/StorageMaterializedView.cpp:508]: We may like to refine the messages to make them brief but with necessary problematic sequence information

File: src/Processors/Streaming/ISource.h (hunk 6 above)
Comment (yl-lisen) [reply src/Processors/Streaming/ISource.h:34]: the `checkpoint/recover` virtual interface is defined in IProcessor

File: src/Storages/Streaming/StorageMaterializedView.cpp (hunk 11 above)
Comment (yl-lisen) [reply src/Storages/Streaming/StorageMaterializedView.cpp]: `const auto &` or `auto &` will get error 
```
error: loop variable '[streaming_source, sn]' binds to a temporary value produced by a range of type 'decltype(zip_view<all_t<const vector<shared_ptr<ISource>, allocator<shared_ptr<ISource>>> &>, all_t<const vector<long long, allocator<long long>> &>>(std::forward<const std::vector<std::shared_ptr<DB::Streaming::ISource>> &>(__rs), std::forward<const std::vector<long long> &>(__rs)))' (aka 'zip_view<ref_view<const std::vector<std::shared_ptr<DB::Streaming::ISource>, std::allocator<std::shared_ptr<DB::Streaming::ISource>>>>, ref_view<const std::vector<long long, std::allocator<long long>>>>') [-Werror,-Wrange-loop-bind-reference]
        for (const auto & [streaming_source, sn] : std::views::zip(streaming_sources, sns))
```
 in fact, `auto` also means a reference , you can see https://quuxplusone.github.io/blog/2020/08/26/wrange-loop-analysis/
 
 or `auto &&`

**Pull request merged**